### DISCUSSION TITLE
Quarkus-OIDC q_session_at and q_session encryption support for Browsers

### DISCUSSION THREAD

**Discussion by doogle-oss** [quarkusio/quarkus#23365] [Q&A]
While using Quarkus-OIDC it seems raw access token and id token is stored in browser cookies. Can that be encrypted to make it more secured. We usually are discouraged to have raw token exposed to the client or store in browser. 
Is there anything to do to make a encryption decryption layer added to the cookie possibly a encryption key to be added to config to do encryption and decryption under the hood.

[a comment by an automated account is not shown]

**Reply by sberyozkin** [reply]
@debu999 We can easily do that with the client secret; I'll open an issue. But note, the JWE encryption will increase the size of the 2 tokens - so if they are large a browser limit might be reached, for 3 tokens it can certainly be a problem but for 2 (since you have no RT) it should not be a problem. However, see also https://quarkus.io/guides/security-openid-connect-web-authentication#session-management (`TokenStateManager`) - you can register a custom one, store them in DB and only keep a DB pointer in a cookie - it has its own challenges like getting the DB/cache state distributed but it is an option which is worth keeping in mind

**Reply by sberyozkin** [reply]
I see you are splitting the cookies so it the cookie size should not be a problem

**Reply by doogle-oss** [reply]
Yes, we split the cookie and keep them separate, any encryption to not expose it would do fine. A simple encryption viz caeser cipher will do as well. It's just not to expose directly on browser. It should be a opt in option. Not sure if anyone would need this unencrypted, it's a risk anyways.

**Reply by sberyozkin** [reply]
@debu999 Sure; we'll follow the JWE process as it is how the tokens if they have to be inner signed first are encrypted, it will be easy with this API: https://quarkus.io/guides/security-jwt-build#encrypt-claims

**Reply by doogle-oss** [reply]
@sberyozkin Thanks for taking this up.

**Reply by sberyozkin** [reply]
@debu999 Np, I'm not looking into it right now as we have a few other issues prioritized but this feature should make it to 2.8.0.CR1 - it would be useful to Quarkus users in general as well when they are concerned about the sensitivity of some claims, without having to write custom `TokenStateManager`

**Reply by doogle-oss** [reply]
Yes. The only intent is not to expose the token in plain sight.

**Reply by doogle-oss** [reply]
Do I need to create a issue to track or it is not required.

**Reply by sberyozkin** [reply]
@debu999 I already did a couple of days back: #23373

**Reply by doogle-oss** [reply]
@sberyozkin thanks. Looking forward to it. You rock as always . Every day i become a bigger fan of quarkus and thanks to lot of efforts put by the team. It's clean and perfect.

**Reply by sberyozkin** [reply]
@debu999 Thanks, np at all, we are here to help, glad you are enjoying Quarkus

**Reply by sberyozkin** [reply]
@debu999, FYI, #23373 has been resolved
